src/python/pants/java/nailgun_io.py

```diff
@@ -50,109 +81,116 @@ def __init__(self, sock, write_handle):
     """Yields the write side of a pipe that will copy appropriately chunked values to the socket."""
-    with _pipe(isatty) as (read_handle, write_handle):
-      writer = NailgunStreamWriter(read_handle, sock, chunk_type, chunk_eof_type,
-                                  buf_size=buf_size, select_timeout=select_timeout)
-      with writer.running():
-        yield write_handle
+    with cls.open_multi(sock, (chunk_type,), chunk_eof_type, isattys=(isatty,)) as ctx:
+      yield ctx
 
+  @classmethod
   @contextmanager
-  def running(self):
-    self.start()
-    try:
-      yield
-    finally:
-      self.stop()
+  def open_multi(cls, sock, chunk_types, chunk_eof_type, isattys, buf_size=None, select_timeout=None):
+    """Yields the write sides of pipes that will copy appropriately chunked values to the socket."""
+    cls._assert_aligned(chunk_types, isattys)
+
+    # N.B. This is purely to permit safe handling of a dynamic number of contextmanagers.
+    with ExitStack() as stack:
+      read_handles, write_handles = zip(
+        # Allocate one pipe pair per chunk type provided.
+        *(stack.enter_context(_pipe(isatty)) for isatty in isattys)
+      )
+      writer = NailgunStreamWriter(
+        read_handles,
+        sock,
+        chunk_types,
+        chunk_eof_type,
+        buf_size=buf_size,
+        select_timeout=select_timeout
+      )
+      with writer.running():
+        yield write_handles, writer
 
   def run(self):
-    while not self.is_stopped:
-      readable, _, errored = select.select([self._in_file], [], [self._in_file], self._select_timeout)
+    while self._in_files and not self.is_stopped:
+      readable, _, errored = select.select(self._in_files, [], self._in_files, self._select_timeout)
 
-      if self._in_file in errored:
-        self.stop()
-        return
+      if readable:
+        for fh in readable:
+          fileno = fh.fileno()
+          data = os.read(fileno, self._buf_size)
 
-      if not self.is_stopped and self._in_file in readable:
-        data = os.read(self._in_file.fileno(), self._buf_size)
-
-        if not self.is_stopped:
-          if data:
-            NailgunProtocol.write_chunk(self._socket, self._chunk_type, data)
-          else:
+          if not data:
+            # We've reached EOF.
             try:
               if self._chunk_eof_type is not None:
                 NailgunProtocol.write_chunk(self._socket, self._chunk_eof_type)
-                self._socket.shutdown(socket.SHUT_WR)  # Shutdown socket sends.
-            except socket.error:  # Can happen if response is quick.
-              pass
             finally:
-              self.stop()
+              try:
+                os.close(fileno)
+              finally:
+                self._in_files.remove(fh)
+          else:
+            NailgunProtocol.write_chunk(
+              self._socket,
+              self._fileno_chunk_type_map[fileno],
+              data
+            )
+
+      if errored:
```

Comment:
Not as familiar with `select` as I should be... does this deserve a log message?